# Popular Python Libraries for Data Analysis & Machine Learning

NumPy – Numerical Python

# Handles arrays, math operations, and numerical computations efficiently.

In [1]:
import numpy as np  # np is a shortcut


# Pandas – Panel Data Analysis

Works with tables (dataframes), easy to filter, group, and analyze data.

In [2]:
import pandas as pd


# Matplotlib – Plotting library

Creates graphs, charts, and plots.

In [3]:
import matplotlib.pyplot as plt  # plt is a shortcut


# Seaborn – Statistical plotting

Makes attractive and easy statistical charts on top of Matplotlib.

In [4]:
import seaborn as sns


# Scikit-learn (sklearn) – Machine learning

Used for building models like regression, classification, clustering.

In [5]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression


# SciPy – Scientific Python

Advanced math, signal processing, optimization, and more.

In [6]:
import scipy


In [7]:
#You can check the version of NumPy installed by running:
print(np.__version__)

2.2.2


In [ ]:
#Running dir(np) in Python will list all attributes, functions,
#  and constants available in the NumPy module. 
# For example:

print(dir(np))


# --- IGNORE ---
#Functions like np.array(), np.mean(), np.reshape()

#Constants like np.pi, np.inf, np.nan

#Classes like np.ndarray, np.matrix

#It’s essentially a full “table of contents” for what NumPy can do.

['False_', 'ScalarType', 'True_', '_CopyMode', '_NoValue', '__NUMPY_SETUP__', '__all__', '__array_api_version__', '__array_namespace_info__', '__builtins__', '__cached__', '__config__', '__dir__', '__doc__', '__expired_attributes__', '__file__', '__former_attrs__', '__future_scalars__', '__getattr__', '__loader__', '__name__', '__numpy_submodules__', '__package__', '__path__', '__spec__', '__version__', '_array_api_info', '_core', '_distributor_init', '_expired_attrs_2_0', '_globals', '_int_extended_msg', '_mat', '_msg', '_pyinstaller_hooks_dir', '_pytesttester', '_specific_msg', '_type_info', '_typing', '_utils', 'abs', 'absolute', 'acos', 'acosh', 'add', 'all', 'allclose', 'amax', 'amin', 'angle', 'any', 'append', 'apply_along_axis', 'apply_over_axes', 'arange', 'arccos', 'arccosh', 'arcsin', 'arcsinh', 'arctan', 'arctan2', 'arctanh', 'argmax', 'argmin', 'argpartition', 'argsort', 'argwhere', 'around', 'array', 'array2string', 'array_equal', 'array_equiv', 'array_repr', 'array_split'

# NumPy Arrays (np.ndarray)

Key Properties:

nD (n-dimensional) – Can be 1D, 2D, 3D, etc.

Homogeneous – All elements must be of the same data type.

Broadcasting allowed – Operations can automatically expand shapes to match.

Vectorization allowed – Apply operations to the entire array without loops.

In [9]:
#Creating NumPy Arrays

#From Python lists (casting)

import numpy as np
a1 = np.array([1, 2, 3, 5, 4, 7, 6], dtype=np.int16)


# Using built-in NumPy functions

In [10]:
np.zeros(5)         # array of zeros
np.ones(5)          # array of ones
np.zeros_like(a1)   # zeros with same shape as a1
np.ones_like(a1)    # ones with same shape as a1
np.full(5, 10)      # array filled with 10
np.arange(0, 10, 2) # range of numbers with step
np.linspace(0, 1, 5) # 5 numbers evenly spaced between 0 and 1
np.random.randint(0, 10, 5) # random integers


array([2, 6, 9, 6, 4], dtype=int32)

# Array Metadata

Data type

In [11]:
print(a1.dtype)  # int16


int16


# Shape

In [12]:
print(a1.shape)  # (7,) → 1D array with 7 elements


(7,)


# Dimensions

In [13]:
print(a1.ndim)   # 1 → 1D


1


# Size

In [14]:
print(a1.size)   # 7 → number of elements


7


# Types of Tables in Data Analysis

Wide table – Headers are in rows

Long table – Headers in columns (most common)

Cross table – Headers in both rows and columns (like pivot tables)

# Checking Array Information

In [15]:
a1.nbytes  
# Shows how much memory (in bytes) the array uses in your system.
# Depends on number of elements + dtype size.


14

In [16]:
a1.size  
# Tells how many elements are present in the array.
# Example: [1,2,3] → size = 3


7

In [17]:
a1.shape  
# Shows the structure (rows, columns).
# 1D example: (7,) means 7 values in a single row.


(7,)

In [18]:
a1.ndim  
# Shows the number of dimensions.
# 1D → 1, 2D → 2, 3D → 3, etc.


1

# Creating Arrays Using NumPy Functions

In [19]:
np.zeros((2,2), dtype=np.int8)
# Creates a 2×2 array filled with 0s.
# Total values = 4 (2 rows × 2 columns).
# dtype=int8 → each value uses 1 byte.


array([[0, 0],
       [0, 0]], dtype=int8)

In [20]:
np.ones((2,3), dtype=np.int16)
# Creates a 2×3 array filled with 1s.
# Total values = 6.
# dtype=int16 → each value uses 2 bytes.


array([[1, 1, 1],
       [1, 1, 1]], dtype=int16)

In [21]:
np.full((3,3), fill_value='abc')
# Creates a 3×3 array where every entry is 'abc'.
# Useful when you need a constant-filled array.


array([['abc', 'abc', 'abc'],
       ['abc', 'abc', 'abc'],
       ['abc', 'abc', 'abc']], dtype='<U3')

In [22]:
np.ones_like(a1)
# Creates a new array with the same shape and dtype as a1,
# but filled with 1s.


array([1, 1, 1, 1, 1, 1, 1], dtype=int16)

In [23]:
np.identity(3)
# Creates a 3×3 identity matrix.
# Main diagonal = 1s, all other positions = 0s.
# Commonly used in linear algebra.


array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [24]:
np.arange(1, 100, 0.5)
# Creates an array starting at 1, ending before 100, increasing by 0.5 each time.
# Useful for fractional or finely spaced ranges when you know the exact step size.
# Example values: 1.0, 1.5, 2.0, 2.5, ..., 99.5


array([ 1. ,  1.5,  2. ,  2.5,  3. ,  3.5,  4. ,  4.5,  5. ,  5.5,  6. ,
        6.5,  7. ,  7.5,  8. ,  8.5,  9. ,  9.5, 10. , 10.5, 11. , 11.5,
       12. , 12.5, 13. , 13.5, 14. , 14.5, 15. , 15.5, 16. , 16.5, 17. ,
       17.5, 18. , 18.5, 19. , 19.5, 20. , 20.5, 21. , 21.5, 22. , 22.5,
       23. , 23.5, 24. , 24.5, 25. , 25.5, 26. , 26.5, 27. , 27.5, 28. ,
       28.5, 29. , 29.5, 30. , 30.5, 31. , 31.5, 32. , 32.5, 33. , 33.5,
       34. , 34.5, 35. , 35.5, 36. , 36.5, 37. , 37.5, 38. , 38.5, 39. ,
       39.5, 40. , 40.5, 41. , 41.5, 42. , 42.5, 43. , 43.5, 44. , 44.5,
       45. , 45.5, 46. , 46.5, 47. , 47.5, 48. , 48.5, 49. , 49.5, 50. ,
       50.5, 51. , 51.5, 52. , 52.5, 53. , 53.5, 54. , 54.5, 55. , 55.5,
       56. , 56.5, 57. , 57.5, 58. , 58.5, 59. , 59.5, 60. , 60.5, 61. ,
       61.5, 62. , 62.5, 63. , 63.5, 64. , 64.5, 65. , 65.5, 66. , 66.5,
       67. , 67.5, 68. , 68.5, 69. , 69.5, 70. , 70.5, 71. , 71.5, 72. ,
       72.5, 73. , 73.5, 74. , 74.5, 75. , 75.5, 76

# np.arange() – You control the step size
# np.arange(start, stop, step)

You decide how much to increase each time.
Good when you know the exact step, like 0.5, 2, 10, etc.
The endpoint is not included.
# Example:
np.arange(1, 10, 0.5)
# Starts at 1, goes up by 0.5, stops before 10.


# np.linspace() – You control the number of values
# np.linspace(start, stop, num)

You specify how many values you want.
The endpoint is included by default.
Useful when you want evenly spaced numbers between two points.
# Example:
# np.linspace(1, 10, 5)
# Creates 5 evenly spaced numbers from 1 to 10 (including 10).
# Output: [ 1. , 3.25, 5.5 , 7.75, 10. ]

In [26]:
np.linspace(1,100,num=50) # This is used when we want a spcified set of numbers

array([  1.        ,   3.02040816,   5.04081633,   7.06122449,
         9.08163265,  11.10204082,  13.12244898,  15.14285714,
        17.16326531,  19.18367347,  21.20408163,  23.2244898 ,
        25.24489796,  27.26530612,  29.28571429,  31.30612245,
        33.32653061,  35.34693878,  37.36734694,  39.3877551 ,
        41.40816327,  43.42857143,  45.44897959,  47.46938776,
        49.48979592,  51.51020408,  53.53061224,  55.55102041,
        57.57142857,  59.59183673,  61.6122449 ,  63.63265306,
        65.65306122,  67.67346939,  69.69387755,  71.71428571,
        73.73469388,  75.75510204,  77.7755102 ,  79.79591837,
        81.81632653,  83.83673469,  85.85714286,  87.87755102,
        89.89795918,  91.91836735,  93.93877551,  95.95918367,
        97.97959184, 100.        ])

In [27]:
np.random.randint(low=1, high=100, size=500)
# Generates 500 random integers.
# Each number will be between 1 (inclusive) and 100 (exclusive).
# In other words, possible values: 1 to 99.
# Useful for simulations, testing, random samples, or creating example datasets.


array([35, 51, 68, 76, 32, 60, 78, 54, 64, 78, 17, 79, 87, 14, 28, 57, 53,
       92, 77, 72, 26, 16, 21, 15, 44, 53, 64, 34, 79, 76, 95, 57, 77, 41,
       93, 42, 69, 82, 44, 40,  2, 36,  1, 55, 47, 76, 39, 49, 47, 32, 53,
        7, 21, 47, 94, 42, 50, 38, 93, 75, 31, 55, 32, 83, 56, 41, 26, 75,
       86, 50, 42, 93, 27, 34, 89, 66,  7, 18, 42,  3, 49, 16, 83, 27, 47,
       80, 92, 47, 98, 83,  2,  3, 78, 76,  7, 53, 40, 27, 84, 67, 72, 90,
       37, 98, 15, 37, 97, 45, 47, 66, 54, 46, 69, 67, 66, 65, 15, 66, 17,
        8, 36, 50, 46, 54, 69, 17, 42, 15, 63, 75, 71, 19, 19, 77,  9, 30,
       33, 57, 36, 83, 64, 95, 35, 44, 55, 27, 81, 43, 91, 62, 30, 48, 37,
       33, 52, 44, 44, 22, 30, 92,  4, 87, 25, 32, 71, 20, 76, 47, 76, 36,
       32, 88,  1, 49, 20, 42, 78, 86, 22, 73, 44, 95, 23, 44, 51, 85, 13,
       68, 73, 74, 99, 81,  1, 91, 27, 20, 94, 88, 96, 39, 41,  5, 87, 13,
       43, 42, 52, 31, 66, 44, 57, 79, 24,  5, 95, 45, 37, 74,  5, 41, 57,
       39, 88, 70, 93, 83

# Random Integers (you already learned this)
# np.random.randint(low=1, high=100, size=500)
 500 random integers from 1 to 99.


# Random Floats (0 to 1)
# np.random.rand(5)
 Creates 5 random floating-point numbers between 0 and 1.
 Example output: [0.23, 0.77, 0.12, 0.98, 0.45]
# np.random.rand(3, 3)
 Creates a 3×3 matrix of random floats (0 to 1).


# Random Floats in Any Range
# np.random.uniform(low=10, high=20, size=5)
 5 random floats between 10 and 20.


# Random Normal Distribution (bell curve)
np.random.normal(loc=0, scale=1, size=10)
# 10 random numbers from a normal distribution.
# loc  = mean
# scale = standard deviation



# Random Choice (pick from a list)
np.random.choice([10, 20, 30, 40], size=5)
# Randomly picks 5 values from the given list.


# Random Matrix of Integers
np.random.randint(0, 10, size=(3, 4))
# Creates a 3×4 matrix of random integers from 0 to 9.


# Set Seed for Reproducible Results
np.random.seed(42)
# Fixes the randomness so the same numbers are generated every time.
# Useful for machine learning experiments.



In [28]:
a2 = np.linspace(1,50,100)

a2.size

100

# Step 1: What does np.linspace(1, 50, 100) do?

Think of this as saying:

“Give me 100 numbers starting at 1 and ending at 50.”

The numbers are evenly spaced.

Example (not exact, just to understand):
1.00
1.50
2.00
2.50
...
50.00
# So the array a2 will contain exactly 100 values.


# Step 2: What does a2.size show? 28

.size = How many values are inside the array.

Since we told NumPy to create 100 values, the output will be:

100



# Very simple summary
| Code                    | Meaning                                |
| ----------------------- | -------------------------------------- |
| `np.linspace(1,50,100)` | Make 100 numbers between 1 and 50      |
| `a2.size`               | Count how many numbers are in a2 → 100 |


In [29]:
a2.reshape((2,5,10))

array([[[ 1.        ,  1.49494949,  1.98989899,  2.48484848,
          2.97979798,  3.47474747,  3.96969697,  4.46464646,
          4.95959596,  5.45454545],
        [ 5.94949495,  6.44444444,  6.93939394,  7.43434343,
          7.92929293,  8.42424242,  8.91919192,  9.41414141,
          9.90909091, 10.4040404 ],
        [10.8989899 , 11.39393939, 11.88888889, 12.38383838,
         12.87878788, 13.37373737, 13.86868687, 14.36363636,
         14.85858586, 15.35353535],
        [15.84848485, 16.34343434, 16.83838384, 17.33333333,
         17.82828283, 18.32323232, 18.81818182, 19.31313131,
         19.80808081, 20.3030303 ],
        [20.7979798 , 21.29292929, 21.78787879, 22.28282828,
         22.77777778, 23.27272727, 23.76767677, 24.26262626,
         24.75757576, 25.25252525]],

       [[25.74747475, 26.24242424, 26.73737374, 27.23232323,
         27.72727273, 28.22222222, 28.71717172, 29.21212121,
         29.70707071, 30.2020202 ],
        [30.6969697 , 31.19191919, 31.68686869, 32

# a2.reshape((2,5,10)) : 29
# First, understand what a2 is

You created:
a2 = np.linspace(1, 50, 100)

This means:

a2 has 100 numbers

They are evenly spaced between 1 and 50

So:   a2.size = 100

# (2, 5, 10) Meaning

2 blocks

Each block has 5 rows

Each row has 10 values

# Visual breakdown:

Block 1 (shape: 5 × 10)
    10 values in row 1
    10 values in row 2
    10 values in row 3
    10 values in row 4
    10 values in row 5

Block 2 (shape: 5 × 10)
    same structure as Block 1

# Because:

2 × 5 × 10 = 100  (matches a2.size)
So reshaping is possible.

# Very simple explanation

reshape((2,5,10)) means:

“Organize these 100 numbers into 2 groups, each containing 5 rows, and each row containing 10 numbers.”

# The values themselves do not change, only their arrangement does.

In [ ]:
a2.reshape((50,2)).T 
# Tanspose could be useful when we are analyzing the info in a cross table

array([[ 1.        ,  1.98989899,  2.97979798,  3.96969697,  4.95959596,
         5.94949495,  6.93939394,  7.92929293,  8.91919192,  9.90909091,
        10.8989899 , 11.88888889, 12.87878788, 13.86868687, 14.85858586,
        15.84848485, 16.83838384, 17.82828283, 18.81818182, 19.80808081,
        20.7979798 , 21.78787879, 22.77777778, 23.76767677, 24.75757576,
        25.74747475, 26.73737374, 27.72727273, 28.71717172, 29.70707071,
        30.6969697 , 31.68686869, 32.67676768, 33.66666667, 34.65656566,
        35.64646465, 36.63636364, 37.62626263, 38.61616162, 39.60606061,
        40.5959596 , 41.58585859, 42.57575758, 43.56565657, 44.55555556,
        45.54545455, 46.53535354, 47.52525253, 48.51515152, 49.50505051],
       [ 1.49494949,  2.48484848,  3.47474747,  4.46464646,  5.45454545,
         6.44444444,  7.43434343,  8.42424242,  9.41414141, 10.4040404 ,
        11.39393939, 12.38383838, 13.37373737, 14.36363636, 15.35353535,
        16.34343434, 17.33333333, 18.32323232, 19.

In [31]:
np.reshape(a2, shape=(50, 2))
# Reshapes the array 'a2' into a new form with 50 rows and 2 columns.
# Since a2 has 100 values, this reshape works because:
# 50 × 2 = 100  → total number of elements must match.
#
# The data inside a2 does NOT change — only the arrangement changes.
#
# After reshaping:
# Row 1 → first 2 values of a2
# Row 2 → next 2 values of a2
# Row 3 → next 2 values
# ...
# Row 50 → last 2 values
#
# Useful for organizing data into a table structure for easier analysis.


array([[ 1.        ,  1.49494949],
       [ 1.98989899,  2.48484848],
       [ 2.97979798,  3.47474747],
       [ 3.96969697,  4.46464646],
       [ 4.95959596,  5.45454545],
       [ 5.94949495,  6.44444444],
       [ 6.93939394,  7.43434343],
       [ 7.92929293,  8.42424242],
       [ 8.91919192,  9.41414141],
       [ 9.90909091, 10.4040404 ],
       [10.8989899 , 11.39393939],
       [11.88888889, 12.38383838],
       [12.87878788, 13.37373737],
       [13.86868687, 14.36363636],
       [14.85858586, 15.35353535],
       [15.84848485, 16.34343434],
       [16.83838384, 17.33333333],
       [17.82828283, 18.32323232],
       [18.81818182, 19.31313131],
       [19.80808081, 20.3030303 ],
       [20.7979798 , 21.29292929],
       [21.78787879, 22.28282828],
       [22.77777778, 23.27272727],
       [23.76767677, 24.26262626],
       [24.75757576, 25.25252525],
       [25.74747475, 26.24242424],
       [26.73737374, 27.23232323],
       [27.72727273, 28.22222222],
       [28.71717172,

In [32]:
import numpy as np

# Suppose 'a2' is a 1D NumPy array with 100 elements
# For example:
a2 = np.arange(100)

# Step 1: Reshape 'a2' into a 2D array with 50 rows and 2 columns
reshaped_array = a2.reshape(50, 2)
# Now 'reshaped_array' looks like this:
# [[ 0,  1],
#  [ 2,  3],
#  [ 4,  5],
#   ...
#  [98, 99]]

# Step 2: Transpose the reshaped array
# Transposing swaps rows and columns
transposed_array = np.transpose(reshaped_array)
# Now 'transposed_array' has 2 rows and 50 columns:
# [[ 0,  2,  4, ..., 96, 98],
#  [ 1,  3,  5, ..., 97, 99]]

# Print the result to see the transformation
print(transposed_array)


[[ 0  2  4  6  8 10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46
  48 50 52 54 56 58 60 62 64 66 68 70 72 74 76 78 80 82 84 86 88 90 92 94
  96 98]
 [ 1  3  5  7  9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47
  49 51 53 55 57 59 61 63 65 67 69 71 73 75 77 79 81 83 85 87 89 91 93 95
  97 99]]


# Slicing or dicing of numpy array: 33

# Summary:

: means "all rows/columns".

start:stop slices a range of rows/columns (stop excluded).

::step selects elements with a specific step, like every other row/column.

In [ ]:
import numpy as np

# Step 1: Create a NumPy array from 1 to 30
a2 = np.arange(1, 31)  # [1, 2, 3, ..., 30]

# Step 2: Reshape it into a 2D array with 5 rows and 6 columns
a2 = a2.reshape((5, 6))
# Now 'a2' looks like this:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# ---------------------------
# Examples of slicing and dicing

# 1. Select the first row
first_row = a2[0, :]
# Output: [1, 2, 3, 4, 5, 6]

# 2. Select the first column
first_col = a2[:, 0]
# Output: [1, 7, 13, 19, 25]

# 3. Select a subarray (rows 1 to 3, columns 2 to 4)
subarray = a2[1:4, 2:5]
# Output:
# [[ 9, 10, 11],
#  [15, 16, 17],
#  [21, 22, 23]]

# 4. Select every other row
every_other_row = a2[::2, :]
# Output:
# [[ 1,  2,  3,  4,  5,  6],
#  [13, 14, 15, 16, 17, 18],
#  [25, 26, 27, 28, 29, 30]]

# 5. Select every other column
every_other_col = a2[:, ::2]
# Output:
# [[ 1,  3,  5],
#  [ 7,  9, 11],
#  [13, 15, 17],
#  [19, 21, 23],
#  [25, 27, 29]]


In [33]:
a2

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
       51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67,
       68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84,
       85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [34]:
import numpy as np

a2 = np.arange(1, 31).reshape((5, 6))
# a2 looks like this:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# Selecting rows 1 to 2 (remember: Python indexing is zero-based, and end index is excluded)
# And selecting column at index 2
result = a2[1:3, 2:3]

print(result)
# Output:
# [[ 9]
#  [15]]


[[ 9]
 [15]]


# Explanation: 34

1:3 → selects rows 1 and 2 (row indices are 1 and 2; row 3 is excluded).

2:3 → selects column 2 only (column indices 2; column 3 is excluded).

The result is a 2D array with shape (2, 1), not a 1D array, because slicing always preserves dimensions.

💡 Tip: If you wanted a 1D array instead of 2D, you could do:
result_1d = a2[1:3, 2]  # This gives [9, 15]

 This way, the column slice 2 returns a flat array instead of a 2D array.

In [35]:

a2 = np.arange(1, 31).reshape((5, 6))
# a2 looks like this:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# Selecting every other row and every other column
result = a2[::2, ::2]

print(result)
# Output:
# [[ 1,  3,  5],
#  [13, 15, 17],
#  [25, 27, 29]]


[[ 1  3  5]
 [13 15 17]
 [25 27 29]]


# Explanation:

::2 in rows → start=0, stop=end, step=2 → selects rows 0, 2, 4.

::2 in columns → start=0, stop=end, step=2 → selects columns 0, 2, 4.

Result is a subset of the original array, picking every other row and column.

✅ Visualized:

Original array (5x6):
[[ 1,  2,  3,  4,  5,  6],
 [ 7,  8,  9, 10, 11, 12],
 [13, 14, 15, 16, 17, 18],
 [19, 20, 21, 22, 23, 24],
 [25, 26, 27, 28, 29, 30]]


# Selected (::2, ::2):
[[ 1,  3,  5],
 [13, 15, 17],
 [25, 27, 29]]


# Tip: You can adjust the step to any number to skip rows/columns differently, e.g., ::3 selects every 3rd row/column.

# Key points: 36
# Broadcasting

Broadcasting means assigning a single value to multiple elements automatically.

Slicing defines the “target area,” and the scalar fills it.

Shape of the target doesn’t need to match the scalar—NumPy automatically expands it.

# Basic Idea

You can perform operations (like +, -, *, /, assignment) on arrays of different shapes.

NumPy automatically aligns shapes and “broadcasts” the smaller array across the larger one.

In [36]:
# Original array
a2 = np.arange(1, 31).reshape((5, 6))
# a2:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# Broadcasting: assign 100 to a subset of the array
a2[1:4, 1::2] = 100

print(a2)


[[  1   2   3   4   5   6]
 [  7 100   9 100  11 100]
 [ 13 100  15 100  17 100]
 [ 19 100  21 100  23 100]
 [ 25  26  27  28  29  30]]


# Step-by-step explanation: 36

1:4 → selects rows 1, 2, 3 (row indices 1 to 3; 4 is excluded).

1::2 → selects columns starting at index 1, every other column (columns 1, 3, 5).

= 100 → broadcasts the scalar 100 to all selected positions.

In [37]:
# Create the array
a2 = np.arange(1, 31).reshape((5, 6))
# Initially:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# ---------------------------
# Step 1: Select a subarray and divide by 50 (broadcasting)
temp_a = a2[1:4, 1::2] / 50
# Subarray selected: rows 1-3, every other column starting at 1
# Shape: (3, 3)
# Broadcasting occurs because 50 (scalar) is applied to all elements
print(temp_a)

# ---------------------------
# Step 2: Add 1 to the subarray (broadcasting)
temp_a = temp_a + 1
# Scalar 1 is broadcast across all elements of temp_a
print(temp_a)

# ---------------------------
# Step 3: Add a ones array to temp_a (vectorization)
temp_a = temp_a + np.ones((3, 3))
# Vectorization: element-wise addition of arrays of same shape
print(temp_a)

# ---------------------------
# Step 4: Assign the modified subarray back to a2 (vectorization)
a2[1:4, 1::2] = temp_a
print(a2)
# Now the selected subarray in a2 has been updated

# ---------------------------
# Step 5: Filtering - find specific values

# Find all values > 5
print(a2[a2 > 5])

# Find all even values
print(a2[a2 % 2 == 0])

# Find even values that are greater than 5
print(a2[(a2 % 2 == 0) & (a2 > 5)])


[[0.16 0.2  0.24]
 [0.28 0.32 0.36]
 [0.4  0.44 0.48]]
[[1.16 1.2  1.24]
 [1.28 1.32 1.36]
 [1.4  1.44 1.48]]
[[2.16 2.2  2.24]
 [2.28 2.32 2.36]
 [2.4  2.44 2.48]]
[[ 1  2  3  4  5  6]
 [ 7  2  9  2 11  2]
 [13  2 15  2 17  2]
 [19  2 21  2 23  2]
 [25 26 27 28 29 30]]
[ 6  7  9 11 13 15 17 19 21 23 25 26 27 28 29 30]
[ 2  4  6  2  2  2  2  2  2  2  2  2 26 28 30]
[ 6 26 28 30]


# Key Concepts: 37

# 1. Broadcasting:

Scalar values like 50 or 1 automatically apply to all elements of an array.

Arrays with compatible shapes can be combined without loops.

# 2. Vectorization:

Operations are applied element-wise across arrays.

Example: temp_a + np.ones((3,3)) adds 1 to every element in temp_a.

# 3. Filtering / Boolean Indexing:

a2[a2 > 5] → selects all elements in a2 that satisfy the condition.

You can combine conditions with & (and) or | (or).

a2[(a2 % 2 == 0) & (a2 > 5)] → selects even values greater than 5.

In [38]:
# Create the array
a2 = np.arange(1, 31).reshape((5, 6))
# a2:
# [[ 1,  2,  3,  4,  5,  6],
#  [ 7,  8,  9, 10, 11, 12],
#  [13, 14, 15, 16, 17, 18],
#  [19, 20, 21, 22, 23, 24],
#  [25, 26, 27, 28, 29, 30]]

# ---------------------------
# Bitwise operators for boolean indexing

# 1. Bitwise AND: &
# Example: fetch even numbers greater than 5
even_gt5 = a2[(a2 % 2 == 0) & (a2 > 5)]
print(even_gt5)

# 2. Bitwise OR: |
# Example: fetch numbers divisible by 3 or 5
div_3_or_5 = a2[(a2 % 3 == 0) | (a2 % 5 == 0)]
print(div_3_or_5)

# 3. Bitwise NOT: ~
# Example: fetch numbers NOT divisible by 2 (odd numbers)
odd_numbers = a2[~(a2 % 2 == 0)]
print(odd_numbers)

# 4. Bitwise XOR: ^
# Example: fetch numbers that are divisible by 2 OR 3, but not both
div_2_xor_3 = a2[(a2 % 2 == 0) ^ (a2 % 3 == 0)]
print(div_2_xor_3)


[ 6  8 10 12 14 16 18 20 22 24 26 28 30]
[ 3  5  6  9 10 12 15 18 20 21 24 25 27 30]
[ 1  3  5  7  9 11 13 15 17 19 21 23 25 27 29]
[ 2  3  4  8  9 10 14 15 16 20 21 22 26 27 28]


# bitwise operators

Bitwise operators are operators that work directly on the binary representation of numbers. Instead of treating numbers as whole integers, they operate bit by bit on the 0s and 1s that make up the number.

# Bitwise Operators in Python / NumPy

Bitwise operators are operators that work **directly on the binary representation of numbers**. They operate **bit by bit** on the 0s and 1s that make up a number.

---

## Common Bitwise Operators

| Operator      | Symbol | Meaning                                             |
|---------------|--------|-----------------------------------------------------|
| AND           | `&`    | 1 if **both bits are 1**, else 0                   |
| OR            | `|`    | 1 if **at least one bit is 1**, else 0            |
| NOT           | `~`    | Inverts bits: 0 → 1, 1 → 0                        |
| XOR           | `^`    | 1 if **bits are different**, 0 if they are the same|
| Left shift    | `<<`   | Shift bits left (multiply by 2 for each shift)     |
| Right shift   | `>>`   | Shift bits right (divide by 2 for each shift)      |

---

## Examples with Numbers

Suppose:

```python
a = 5  # binary 0101
b = 3  # binary 0011
